lightning/src/util/persist.rs

```diff
@@ -250,3 +261,942 @@ where
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			key,
+			&monitor_bytes,
+		) {
+			Ok(_) => {
+				if let Some(old_update_id) = self
+					.last_monitor_update_id
+					.borrow_mut()
+					.insert(monitor_name, monitor.get_latest_update_id())
+				{
+					// Typically, we'll clean up between the last two known full monitors
+					let start = old_update_id;
+					let mut end = monitor.get_latest_update_id();
+					if monitor.get_latest_update_id() == CLOSED_CHANNEL_UPDATE_ID {
+						// We don't want to clean the rest of u64, so just do possible pending updates
+						end = old_update_id + self.maximum_pending_updates;
+					}
+					if let Err(e) =
+						self.lazy_delete_updates(&MonitorName::from(funding_txo), start, end)
+					{
+						log_error!(
+							self.logger,
+							"error cleaning up channel monitor updates {}",
+							e
+						);
+					};
+				};
+				chain::ChannelMonitorUpdateStatus::Completed
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"error writing channel monitor {}",
+					e
+				);
+				chain::ChannelMonitorUpdateStatus::PermanentFailure
+			},
+		}
+	}
+	
+
+	/// Persists a channel update, writing only the update to the provided
+	/// [`KVStore`] if possible.
+	///
+	/// In some cases, this will forward to [`MonitorUpdatingPersister::persist_new_channel`]:
+	///   
+	///   - The number of pending updates exceeds
+	///   [`self.maximum_pending_updates`]
+	///   - LDK will command re-persisting the entire monitor through this
+	/// function, specifically when `update` is `None`.
+	///   - If the update is at [`CLOSED_CHANNEL_UPDATE_ID`]
+	///   - If we don't know the last time a full monitor was persisted (edge
+	///   case)
+	fn update_persisted_channel(
+		&self,
+		funding_txo: OutPoint,
+		update: Option<&ChannelMonitorUpdate>,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		update_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		if let Some(update) = update {
+			if update.update_id != CLOSED_CHANNEL_UPDATE_ID {
+				if let Some(last_monitor_update_id) = self
+					.last_monitor_update_id
+					.borrow()
+					.get(&MonitorName::from(funding_txo))
+				{
+					if update.update_id - last_monitor_update_id < self.maximum_pending_updates {
+						// This is an update to the monitor, which we persist to
+						// be applied on restart. 
+						//
+						// IMPORTANT: update_id: [`MonitorUpdateId`] is not to
+						// be confused with [`ChannelMonitorUpdate`]'s
+						// update_id.
+						let ns = match MonitorName::try_from(funding_txo) {
+							Ok(monitor_name) => self.monitor_update_namespace(&monitor_name),
+							Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+						};
+						let update_name = match UpdateName::try_from(update.update_id) {
+							Ok(update_name) => update_name,
+							Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+						};
+						let key = update_name.as_str();
+						return match self.kv_store.write(
```

Comment:
I thought I left a note about this somewhere -- I tried, but it's tricky because of the multiple `if lets` here, and I think they're necessary. I needed to use the unstable `#![feature(let_chains)]`, which seemed like a non-starter. So I have these four `if`s, and I wasn't keen to put the forward in an else on each of them, right? 

The only way I think I can do it is collect all the conditions to go/no-go the update write in 3-4 bindings (it depends whether we end up keeping the hashmap). Then write `if all([a,b,c,d]) {} else {}`. Is that preferable?

Reply:
After getting rid of the hashmap, this was easier to refactor without the early return, so it should be fixed now.